qtt/algorithms/markov_chain.py

```diff
@@ -0,0 +1,241 @@
+        self._validity_check()
+
+    def _validity_check(self):
+        assert(len(self.states) == len(self.jump_chain))
+        if not (np.allclose(np.sum(self.jump_chain, axis=0), 1)):
+            raise Exception('jump chain matrix should represent probabilities')
+
+        assert(np.all(self.holding_parameters > 0))
+
+    @staticmethod
+    def _create_generator_matrix(holding_parameters, jump_chain):
+        G = np.array(jump_chain, copy=True)
+        for ii in range(G.shape[0]):
+            G[:, ii] = holding_parameters[ii] * jump_chain[:, ii]
+        for ii in range(G.shape[0]):
+            G[ii, ii] = -holding_parameters[ii]
+        return G
+
+    def number_of_states(self):
+        """ Return the number of states in the model """
+        return len(self.states)
+
+    def transition_matrix(self, delta_time):
+        """ Return the transition matrix for a specified amount of time """
+        P = scipy.linalg.expm(delta_time * self.generator_matrix)
+        return P
+
+    def __repr__(self):
+        return "%s(id=0x%x, states=%s, generator=%s)" % (self.__class__.__name__,
+                                                         id(self), self.states, self.generator_matrix)
+
+    def stationary_distribution_direct(self):
+        """ Return the stationary distrubution of the model
+
+        From https://www.probabilitycourse.com/chapter11/11_3_2_stationary_and_limiting_distributions.php, Theorem 11.3
+        """
+        pi_tilde = self.stationary_distribution_discrete(self.jump_chain)
+        norm = np.sum((pi_tilde / self.holding_parameters))
+        pi = (pi_tilde / self.holding_parameters) / norm
+        return pi
+
+    def stationary_distribution(self):
+        """ Return the stationary distrubution of the model
+
+        From https://vknight.org/unpeudemath/code/2015/08/01/simulating_continuous_markov_chains.html
+        """
+        Q = self.generator_matrix
+        n = Q.shape[0]
+        A = np.vstack((Q, np.ones((1, n))))
+        B = np.zeros((n + 1, 1))
+        B[-1] = 1
+        pi = _solve_least_squares(A, B)
+        return pi
+
+    @staticmethod
+    def stationary_distribution_discrete(jump_chain):
+        """ Return the stationary distrubution for a Markov chain """
+
+        n = jump_chain.shape[0]
+        A = np.vstack((jump_chain - np.eye(n), np.ones((1, n))))
+        B = np.zeros((n + 1, 1))
+        B[-1] = 1
+        pi = _solve_least_squares(A, B)
+        return pi
+
+    def generate_sequence(self, length, delta_time, initial_state=None):
+        """ Generate a random sequence with the model
+
+        Args:
+            length (int): number of elements in the sequence
+            delta_time (float): time step to be used
+            initial_state (None or int or list): If an int, then use that state is initial state. If None then take a random state weighted by the stationary distribution        
```

Comment:
It is not clear to me what initial_state here means. Does it mean that before hand you decide how many of the traces are spin-up and how many are spin-down?